# hw 507
請比較Learning Regression及CART regression Tree模型預測boston房價的效能。
- 計算模型的訓練資料集與測試資料集的$R^2$
- 比較Linear Regression與CART regression Tree模型的測試資料集的效能($R^2$)

In [1]:
#導入所需的套件
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import r2_score

In [2]:
#讀取boston資料
import kagglehub

# Download latest version
path = kagglehub.dataset_download("altavish/boston-housing-dataset")

# download dataset
df = pd.read_csv(f"{path}/HousingData.csv")
df.head()

d:\Anaconda\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


,CRIM,ZN,INDUS,CHAS,NOX,RM,AGE,DIS,RAD,TAX,PTRATIO,B,LSTAT,MEDV
0,0.00632,18.0,2.31,0.0,0.538,6.575,65.2,4.0900,1,296,15.3,396.90,4.98,24.0
1,0.02731,0.0,7.07,0.0,0.469,6.421,78.9,4.9671,2,242,17.8,396.90,9.14,21.6
2,0.02729,0.0,7.07,0.0,0.469,7.185,61.1,4.9671,2,242,17.8,392.83,4.03,34.7
3,0.03237,0.0,2.18,0.0,0.458,6.998,45.8,6.0622,3,222,18.7,394.63,2.94,33.4
4,0.06905,0.0,2.18,0.0,0.458,7.147,54.2,6.0622,3,222,18.7,396.90,NaN,36.2


# 作業開始

### 資料先處理

In [3]:
# 檢查缺失值
print("缺失值統計：")
print(df.isnull().sum())

# 補齊缺失值 (使用中位數)
df = df.fillna(df.median())

# 分離特徵與目標變數
X = df.drop('MEDV', axis=1)
y = df['MEDV']

# 切分訓練集與測試集 (80/20)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
print('\n資料切分完成')

缺失值統計：
CRIM       20
ZN         20
INDUS      20
CHAS       20
NOX         0
RM          0
AGE        20
DIS         0
RAD         0
TAX         0
PTRATIO     0
B           0
LSTAT      20
MEDV        0
dtype: int64

資料切分完成


### 建立模型

In [4]:
# 1. Linear Regression
lr_model = LinearRegression()
lr_model.fit(X_train, y_train)

# 2. CART Regression Tree
cart_model = DecisionTreeRegressor(random_state=42)
cart_model.fit(X_train, y_train)

print('模型訓練完成')

模型訓練完成


### 效能評量

In [5]:
# 計算 Linear Regression 的 R2
lr_train_r2 = r2_score(y_train, lr_model.predict(X_train))
lr_test_r2 = r2_score(y_test, lr_model.predict(X_test))

# 計算 CART Regression Tree 的 R2
cart_train_r2 = r2_score(y_train, cart_model.predict(X_train))
cart_test_r2 = r2_score(y_test, cart_model.predict(X_test))

print(f"Linear Regression - Train R2: {lr_train_r2:.4f}, Test R2: {lr_test_r2:.4f}")
print(f"CART Regression Tree - Train R2: {cart_train_r2:.4f}, Test R2: {cart_test_r2:.4f}")

Linear Regression - Train R2: 0.7426, Test R2: 0.6591
CART Regression Tree - Train R2: 1.0000, Test R2: 0.6810


# 結論

根據實驗結果：
1. **Linear Regression**: 訓練集 $R^2$ 約為 0.7426，測試集 $R^2$ 約為 0.6591。
2. **CART Regression Tree**: 訓練集 $R^2$ 為 1.0000 (完美擬合)，測試集 $R^2$ 約為 0.6810。

**比較與分析**：
- **預測效能**：在測試資料集上，**CART Regression Tree (0.6810)** 的表現優於 **Linear Regression (0.6591)**。
- **擬合情況**：CART 模型在訓練集達到 1.0000，顯示出明顯的過擬合 (Overfitting) 現象，但在本案例中其泛化能力仍略強於線性回歸。